pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +258,275 @@ func (p *Pod) Skip() bool {
 	return false
 }
 
+func (p *Pod) groupName() string {
+	return p.Object().GetLabels()[GroupNameLabel]
+}
+
+func (p *Pod) groupTotalCount() (int, error) {
+	if p.groupName() == "" {
+		return 0, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+
+	gtcAnnotation, ok := p.Object().GetAnnotations()[GroupTotalCountAnnotation]
+	if !ok {
+		return 0, fmt.Errorf("failed to extract '%s' annotation",
+			GroupTotalCountAnnotation)
+	}
+
+	gtc, err := strconv.Atoi(gtcAnnotation)
+	if err != nil {
+		return 0, err
+	}
+
+	if gtc < 1 {
+		return 0, fmt.Errorf("incorrect annotation value '%s=%s': group total count should be greater than zero",
+			GroupTotalCountAnnotation, gtcAnnotation)
+	}
+
+	return gtc, nil
+}
+
+func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet, error) {
+	groupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return nil, err
+	}
+
+	var resultPodSets []kueue.PodSet
+
+	for _, podInGroup := range podsInGroup.Items {
```

Comment:
we need to ignore the Pods with phase Failed.

Reply:
Actually, we can't fully ignore the Pods with phase Failed, otherwise we would think that the Workload object no longer matches.

One possible workaround is that we ignore the Failed objects, but we can change the logic for `equivalentToWorkload` a little, to allow for number of pods lower than the saved counter in the Workload object.

Reply:
Changed this behavior in 09c09ceb